Add support for ST LIS3MDL MEMS sensor

bforbort opened the pull request:
This PR addresses iNavFlight/inav#2607 

The [ST LIS3MDL](https://www.st.com/en/mems-and-sensors/lis3mdl.html) 3-axis MEMS magnetic field sensor is coming equipped on some newer GPS receivers.  

It offers an extremely small footprint and user-selectable scales of ±4/±8/±12/±16 gauss.

Testing with [mRo uGPS ublox SAM M8Q](https://store.mrobotics.io/product-p/mro-ugps-samm8q-01.htm) on a Theer wing.

bforbort committed:
Add support for ST LIS3MDL 3-axis MEMS magnetic field sensor

This sensor is coming equipped on some newer GPS receivers.

Testing with mRo uGPS ublox SAM M8Q on a Theer wing.
https://store.mrobotics.io/product-p/mro-ugps-samm8q-01.htm

bforbort commented on src/main/target/common_hardware.c:
```diff
@@ -162,6 +162,14 @@
+        #define LIS3MDL_I2C_BUS MAG_I2C_BUS
+    #endif
+    BUSDEV_REGISTER_I2C(busdev_lis3mdl,     DEVHW_LIS3MDL,      LIS3MDL_I2C_BUS,    0x1E,               NONE,           DEVFLAGS_NONE);
+//  BUSDEV_REGISTER_I2C(busdev_lis3mdl,     DEVHW_LIS3MDL,      LIS3MDL_I2C_BUS,    0x1C,               NONE,           DEVFLAGS_NONE);
```
If the SDO/SA1 pin of the LIS3MDL is connected to the voltage supply, the address is 0x1E, otherwise, if the SDO/SA1 pin is
connected to ground, the address is 0x1C.

bforbort commented on src/main/drivers/compass/compass_lis3mdl.c:
```diff
@@ -0,0 +1,203 @@
+    return true;
+}
+
+bool lis3mdlDetect(magDev_t * mag)
```
The I2C address of the LIS3MDL can be changed by connecting it's SDO/SA1 to either supply or ground.  As there is no way to know which configuration is used, it may be necessary to check both addresses.

Is it possible to change the I2C address of a device here, and then re-run the detection, or can that only be done in common_hardware.c?

digitalentity replied:
I'm planning to add support for this

bforbort replied:
As I'm porting this over to Betaflight, I noticed that they have already added this ability. However, knowing that PX4 only supports the 0x1E, and since Betaflight is already having major space issues, I probably won't bother checking for the 0x1C address.

bforbort committed:
Detect LIS3MDL in either address configuration

bforbort commented on src/main/target/common_hardware.c:
```diff
@@ -167,7 +167,7 @@
     #endif
     BUSDEV_REGISTER_I2C(busdev_lis3mdl,     DEVHW_LIS3MDL,      LIS3MDL_I2C_BUS,    0x1E,               NONE,           DEVFLAGS_NONE);
-//  BUSDEV_REGISTER_I2C(busdev_lis3mdl,     DEVHW_LIS3MDL,      LIS3MDL_I2C_BUS,    0x1C,               NONE,           DEVFLAGS_NONE);
+    BUSDEV_REGISTER_I2C(busdev_lis3mdl_a2,  DEVHW_LIS3MDL_A2,   LIS3MDL_I2C_BUS,    0x1C,               NONE,           DEVFLAGS_NONE);
```
Creating a second bus device works fine, the mag is detected with either address.  It just doesn't seem very graceful.  Any other ideas?

bforbort committed:
Revert "Detect LIS3MDL in either address configuration"

This reverts commit d7addd5cc03a75667d65ada250c06f06492f9783.

digitalentity commented on src/main/drivers/bus.h:
```diff
@@ -102,6 +102,8 @@ typedef enum {
     DEVHW_QMC5883,
     DEVHW_MAG3110,
+    DEVHW_LIS3MDL,
+    DEVHW_LIS3MDL_A2,
```
Having two items for the same hardware is not wise. I suggest keeping the one that is compatible with off the shelf modules for Pixhawk.

digitalentity replied:
I'm planning to address multiple possible I2C addresses for the same sensor chip.

bforbort replied:
Looks like PX4 is only supporting the address 0x1E.  Maybe it's not even necessary to check for the alternative address?

https://github.com/PX4/Firmware/blob/02d4405a6205cfa050925ac0edcf4d0a13382632/src/drivers/magnetometer/lis3mdl/lis3mdl_i2c.cpp#L62

https://github.com/PX4/Firmware/blob/02d4405a6205cfa050925ac0edcf4d0a13382632/src/drivers/magnetometer/lis3mdl/lis3mdl_i2c.cpp#L88-L92

bforbort committed:
Add comment about second I2C address

bforbort committed:
Cleanup

Removed unused includes
Reduced detection retry attempts
Removed delays from init

Bench tested, detected and all init messages where acknoloadged
